src/Lean/Elab/PreDefinition/PartialFixpoint/Main.lean

```diff
@@ -72,47 +70,67 @@ def mkMonoPProd (hmono₁ hmono₂ : Expr) : MetaM Expr := do
   mkAppOptM ``PProd.monotone_mk #[none, none, none, inst₁, inst₂, inst, none, none, hmono₁, hmono₂]
 
 def partialFixpoint (preDefs : Array PreDefinition) : TermElabM Unit := do
-  -- We expect all functions in the clique to have `partial_fixpoint` syntax
+  -- We expect all functions in the clique to have `partial_fixpoint` or `greatest_fixpoint` syntax
   let hints := preDefs.filterMap (·.termination.partialFixpoint?)
   assert! preDefs.size = hints.size
+  -- We check if any fixpoints were defined lattice-theoretically
+  let isLattice := hints.any fun x => isLatticeTheoretic x.fixpointType
+  -- If yes, then we expect all of them to be defined using lattice theory
+  if isLattice then
+    unless hints.all fun x => isLatticeTheoretic x.fixpointType do
+      throwError "all functions in the clique must be defined using lattice theory, but some are not"
```

Comment:
As hinted at in the comment for the test file: This check could happen earlier, in `checkTerminationByHints`, as for the other termination hints. Then this could become an `assert!` here.